File: JS - Fundamentals/09. Associative Arrays - Exercise/legendaryFarming.js
function solve(input) {
    const array = input.split(' ');
    let storage = { fragments: 0, shards: 0, motes: 0 };
    let junk = {};
    let objective = 250;
    const result = [];
    
    for (let i = 0; i < array.length; i++) {
        let quantity = Number(array[i]);
        i++;
        let material = array[i].toLowerCase();
        
        if(material === 'fragments' || material === 'shards' || material === 'motes'){
            storage[material] += quantity;
        } else {
            if(!junk.hasOwnProperty(material)){
                junk[material] = quantity;
            } else {
                junk[material] += quantity;
            }
        }

        if (storage.shards >= objective) {
            result.push('Shadowmourne obtained!');
            storage.shards -= objective;
            break;
        } else if (storage.fragments >= objective) {
            result.push('Valanyr obtained!');
            storage.fragments -= objective;
            break;
        } else if (storage.motes >= objective) {
            result.push('Dragonwrath obtained!');
            storage.motes -= objective;
            break;
        }
    }

    Object.entries(storage)
    .sort((a,b) => b[1] - a[1] || a[0].localeCompare(b[0]))
    .forEach(element =>{
        result.push(`${element[0]}: ${element[1]}`)
    });

    Object.entries(junk)
    .sort((a,b) => a[0].localeCompare(b[0]))
    .forEach(element => {
        result.push(`${element[0]}: ${element[1]}`)
    });

    return result.join('\n');
}

console.log(solve(
    '3 Motes 5 stones 5 Shards 6 leathers 255 fragments 7 Shards'
));

File: Java Fundamentals/08. Maps, Lambda and Stream API - Exercise/src/LegendaryFarming.java
import java.util.*;

public class LegendaryFarming {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        Map<String, Integer> collectedResources = new TreeMap<>();
        Map<String, Integer> collectedJunk = new TreeMap<>();

        collectedResources.put("fragments", 0);
        collectedResources.put("shards", 0);
        collectedResources.put("motes", 0);

        String winner = "";

        while(winner.equals("")){
            String[] currentResources = scan.nextLine().split("\\s+");

            winner = populateReserves(collectedResources, collectedJunk, currentResources, winner);
        }

        printOutput(collectedResources, collectedJunk, winner);

    }

    public static String populateReserves(Map<String, Integer> recourses, Map<String, Integer> junk, String [] currentResources, String winner){
        for (int i = 0; i < currentResources.length -1; i++) {
            int quantity = Integer.parseInt(currentResources[i]);
            i++;
            String recourseType = currentResources[i].toLowerCase(Locale.ROOT);

            if(recourseType.equals("fragments") || recourseType.equals("shards") || recourseType.equals("motes")){
                recourses.put(recourseType, recourses.get(recourseType) + quantity);
                winner = findWinner(recourses, recourseType, winner);

                if(winner.equals("")==false){
                    break;
                }
            } else {
                junk.putIfAbsent(recourseType, 0);
                junk.put(recourseType, junk.get(recourseType) + quantity);
            }
        }

        return winner;
    }

    public static String findWinner(Map<String, Integer> recourses, String item, String winner){
        if(recourses.get(item) >= 250){
            switch (item){
                case "shards": winner = "Shadowmourne"; break;
                case "fragments": winner = "Valanyr"; break;
                case "motes": winner = "Dragonwrath"; break;
                default: break;
            }

            recourses.put(item, recourses.get(item) - 250);
        }

        return winner;
    }

    public static void printOutput(Map<String, Integer>collectedResources, Map<String, Integer> collectedJunk, String winner){
        System.out.printf("%s obtained!\n", winner);

        collectedResources
                .entrySet()
                .stream()
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed().thenComparing(Map.Entry.comparingByKey()))
                .forEach(result -> System.out.printf("%s: %d%n", result.getKey(), result.getValue()));


        collectedJunk
                .entrySet()
                .stream()
                .sorted(Map.Entry.<String, Integer>comparingByKey())
                .forEach(result -> System.out.printf("%s: %d%n", result.getKey(), result.getValue()));
    }
}
